### ISSUE TITLE
BUG: gpu_predictor in Python

### ISSUE THREAD

**Issue by dangoldner** [dmlc/xgboost#3162]
Hello,
I am getting similar behavior to #2965, but in Python. 

## Environment info
Operating System: linux (aws p2.xlarge)

Compiler: gcc

Package used (python/R/jvm/C++): python

`xgboost` version used: 0.7

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`) - a1b48afa41f888

2. Logs will be helpful (If logs are large, please upload as attachment).

If you are using jvm package, please 

1. add [jvm-packages] in the title to make it quickly be identified
2. the gcc version and distribution

If you are using python package, please provide

1. The python version and distribution: Python 3.6.4 Anaconda (default Jan 16 2018, 18:10:19)

2. The command to install `xgboost` if you are not installing from source - Installed from source

If you are using R package, please provide

1. The R `sessionInfo()`
2. The command to install `xgboost` if you are not installing from source

## Steps to reproduce

```
gpu_params = {'tree_method':'gpu_hist', 
              'predictor':'cpu_predictor',
              'n_jobs': -1,
              'seed': 123
             }
m = xgb.XGBRegressor(**gpu_params).fit(X_ts, y_ts)
m.score(X_ts, y_ts) 
m.score(X_test, y_test)
```

Then do exactly the same thing changing only "cpu_predictor" to "gpu_predictor". Test score changes dramatically, predicted values change dramatically. With cpu, the values I get are sensible and verified.

**Comment by RAMitchell** [comment]
@dangoldner I need your complete script including dataset to reproduce this.

**Comment by dangoldner** [comment]
That's fair. Data are proprietary-I'll see if I can cook up some data that do the same thing.

**Comment by dangoldner** [comment]
Still happening for me but I can't find a vanilla reproduction. I'll reopen if I can find a good example.

**Comment by dangoldner** [comment]
Ok got it: 

Complete script is at https://gist.github.com/38a8c1d16d26283c9e7759f06be0fcd5. 

It seems to be related to the size of the problem: If `m, n = 10000, 7` and `tr_size = 6000`, the bug does _not_ occur. With larger sizes as shown in the gist, it does. 

Also note that the problem appears starting with the _second_ prediction, not the first.

**Comment by dangoldner** [comment]
I pulled and rebuilt xgboost, but the behavior persists. 

Commit hash 00d9728

I ran the same script (below), from the terminal: 
![screen shot 2018-03-21 at 1 20 57 pm](https://user-images.githubusercontent.com/26610553/37729348-c2009310-2d0a-11e8-9fdb-af7dd4d26cf9.png)

```
import numpy as np
import xgboost as xgb

# *Note:* If `m, n = 10000, 7` and `tr_size = 6000`, the bug does _not_ occur.

# Data are arbitrary but structured so good predictions are possible.
m, n = 1500000, 14
tr_size = 250000
X = np.random.rand(m,n)
y = 200*np.matmul(X,np.arange(-3,-3+n))
X_train, y_train = X[:tr_size,:], y[:tr_size]
X_test, y_test = X[tr_size:,:], y[tr_size:]

# First with cpu_predictor
gpu_params = {'tree_method':'gpu_hist', 
             'predictor':'cpu_predictor',
             'n_jobs': -1,
             'seed': 123
            }
m = xgb.XGBRegressor(**gpu_params).fit(X_train, y_train)
print(m)
print(m.score(X_train, y_train))
print(m.score(X_test,y_test))

# Now with gpu_predictor
gpu_params = {'tree_method':'gpu_hist', 
              'predictor':'gpu_predictor',
              'n_jobs': -1,
              'seed': 123
             }

m = xgb.XGBRegressor(**gpu_params).fit(X_train, y_train)
print(m)
print(m.score(X_train, y_train))
print(m.score(X_test, y_test))
```

**Comment by RAMitchell** [comment]
Ok I will have another look. A problem I am having in debugging this is that the problem goes away sometimes.

**Comment by dangoldner** [comment]
Thanks!

**Comment by dangoldner** [comment]
Seems to have fixed it on my system -- see note in #3217
